24H1KR3: Developer Experience (#1174)
vtexdocs/dev-portal-content: docs/guides/VTEX-Platform-Overview/developer-experience.md

---
title: "Developer experience"
slug: "developer-experience"
hidden: false
createdAt: "2024-05-24T10:18:55.338Z"
updatedAt: "2024-05-24T10:18:55.338Z"
excerpt: "Learn how our platform enhances the developer experience."
---

> This guide provides an overview of the developer experience on the VTEX platform. In the following sections, you will understand how to enhance your experience on the platform and maximize its potential.

Reviewer: ```suggestion
This guide offers an overview of the tools and resources available to development teams. In the following sections, you will learn about the various features we provide for developers in our ecosystem and the opportunities for sharing knowledge and collaborating within the community.
```

- [App development](#app-development): Develop and manage your VTEX IO apps.
- [Storefront development](#storefront-development): Explore VTEX’s storefront solutions.
- [Contributions and collaboration](#contributions-and-collaboration): Contribute to the VTEX ecosystem by sharing your knowledge.

## App development

> [VTEX IO](https://developers.vtex.com/docs/guides/vtex-io-documentation-what-is-vtex-io) is a low-code development platform that simplifies building, managing, and scaling ecommerce solutions by offering a comprehensive suite of tools and services,  manual coding efforts while ensuring performance, scalability, and security for your projects.

Reviewer: I think we avoid using low-code because it's not very accurate

A Command Line Interface (CLI) is a text-based interface used to interact with a computer system or software application by typing commands into a terminal or command prompt. It allows users to perform tasks such as running programs, managing files, and configuring settings efficiently.

To build and develop any project in the VTEX IO, use the pre-configured [**VTEX IO CLI**](https://developers.vtex.com/docs/guides/vtex-io-documentation-vtex-io-cli-installation-and-command-reference). This CLI is the starting and ending point for any development on the VTEX platform. It provides the ability to perform actions relevant to the development process, such as releasing new app versions, for example.

>ℹ️ For [FastStore](https://developers.vtex.com/docs/guides/faststore/docs-what-is-faststore) projects, we also offer the [**FastStore CLI**](https://developers.vtex.com/docs/guides/faststore/getting-started-3-faststore-cli). This CLI tool connects the source code in [`@faststore/core`](https://developers.vtex.com/docs/guides/faststore/project-structure-overview#packagejson) package with your customizations and extensions for your FastStore project.

#### Workspaces

> [Workspaces](https://developers.vtex.com/docs/guides/vtex-io-documentation-workspace) are isolated environments for app development, acting as separate versions of the same account.
> 
> There are three types of workspaces in VTEX:

Reviewer: ```suggestion
[Workspaces](https://developers.vtex.com/docs/guides/vtex-io-documentation-workspace) are isolated environments for app development, acting as separate versions of the same account. There are three types of workspaces in VTEX:
```

- **Master**: A unique production workspace that reflects the live store content.

> - [**Development workspaces**](https://developers.vtex.com/docs/guides/vtex-io-documentation-creating-a-development-workspace): Provide you freedom in code development without affecting user traffic and the live store. You must use a development workspace when making changes or developing a new VTEX IO app.

Reviewer: ```suggestion
- [**Development workspaces**](https://developers.vtex.com/docs/guides/vtex-io-documentation-creating-a-development-workspace): Environments that allow testing changes in real time. They provide freedom in code development without affecting user traffic and the live store. Necessary when making changes to storefronts or developing a new VTEX IO app.
```

- [**Production workspaces**](https://developers.vtex.com/docs/guides/vtex-io-documentation-creating-a-production-workspace): Mainly used to validate VTEX IO apps. These workspaces handle traffic, so they can be used to run [A/B tests](#ab-tests) and can be [promoted to master](https://developers.vtex.com/docs/guides/vtex-io-documentation-promoting-a-workspace-to-master).

To learn how to manage workspaces effectively in your projects, check out the [Best practices on workspaces management](https://developers.vtex.com/docs/guides/vtex-io-documentation-workspaces-best-practices) guide.

#### A/B Tests

> After [linking your app](https://developers.vtex.com/docs/guides/vtex-io-documentation-linking-an-app) in a production workspace, run an [A/B test](https://developers.vtex.com/docs/guides/vtex-io-documentation-running-native-ab-testing) to compare traffic and choose the workspace that performs better in terms of user engagement and conversions.

Reviewer: ```suggestion
After [linking your app](https://developers.vtex.com/docs/guides/vtex-io-documentation-linking-an-app) in a production workspace, it is possible to run an [A/B test](https://developers.vtex.com/docs/guides/vtex-io-documentation-running-native-ab-testing) to compare traffic and choose the workspace that performs better in terms of user engagement and conversions.
```

By leveraging builders, developers can extend the functionality of the VTEX platform without starting from scratch. This modular approach promotes flexibility, scalability, and efficiency in developing and customizing VTEX-based ecommerce solutions, accelerating development cycles, promoting code reusability, and facilitating rapid [deployment](#deployment) of new functionalities.

#### Edition apps

[Edition apps](https://developers.vtex.com/docs/guides/vtex-io-documentation-edition-app) are collections of VTEX apps that define the functionalities available in a VTEX account. They consist of a bundle of settings, policies, backend, and frontend apps, aiming to streamline the setup and management of essential configurations quickly and efficiently.

We offer two native Edition Apps:

- **Edition Business**: For Legacy CMS Portal stores.

> - **Edition Store**: For Store Framework stores.

Reviewer: Edition Store should probably come first on the list

We provide some ready-to-use [code templates](https://developers.vtex.com/docs/guides/code-samples) to simplify your development process when creating a VTEX IO app. Using one of them, you do not need to write the boilerplate code, ensuring development agility, quality and compliance with platform standards.

>ℹ️ To learn about the relevant aspects to consider when developing VTEX IO apps, check out the [Before creating an app for your store](https://help.vtex.com/tracks/vtex-store-overview--eSDNk26pdvemF3XKM0nK9/7euXDZR5CCnVFSrXyczIhu#before-creating-an-app-for-your-store) article.

##### **VTEX App Store**

You can make your apps available to the VTEX ecosystem on the [VTEX App Store](https://developers.vtex.com/docs/guides/vtex-app-store), a marketplace for VTEX IO plug-and-play solutions.

>ℹ️ To learn about the homologation requirements for the VTEX App Store, check out the [App Store Guidelines](https://developers.vtex.com/docs/guides/vtex-io-documentation-homologation-requirements-for-vtex-app-store).

#### Deployment

> Once your new app has been developed, it must be deployed to the VTEX IO production workspace to make it publicly available.
> 
> To [deploy a new app version](https://developers.vtex.com/docs/guides/vtex-io-documentation-making-your-new-app-version-publicly-available), follow these steps:

Reviewer: ```suggestion
Once a new app has been developed, it must be deployed to the VTEX IO production workspace to make it publicly available. The deployment process is the following:
```

![image](https://cdn.jsdelivr.net/gh/vtexdocs/dev-portal-content@main/docs/guides/VTEX-Platform-Overview/making-an-app-publicly-available.png)

- [**Releasing a new app version**](https://developers.vtex.com/docs/guides/vtex-io-documentation-releasing-a-new-app-version): Marks the beginning by releasing the app using Git, updating the version in `manifest.json`, and documenting changes in CHANGELOG.md.
- [**Publishing**](https://developers.vtex.com/docs/guides/vtex-io-documentation-publishing-an-app): The app version becomes a candidate version, ready for testing and validation in a production workspace.

> - [**Testing**](https://developers.vtex.com/docs/guides/ab-tests): Run [A/B tests](#ab-tests) and others, such as performance and security testing.

Reviewer: If we don't have links with more information for performance and security testing, maybe this info should be removed as it might cause confusion

- [**Deploying**](https://developers.vtex.com/docs/guides/vtex-io-documentation-deploying-the-app-stable-version): Update the app version across all accounts where the app is installed.

> - [**Promotiong**](https://developers.vtex.com/docs/guides/vtex-io-documentation-promoting-a-workspace-to-master): Optional step to promote the production workspace to master.
> 

Reviewer: ```suggestion
- [**Promoting**](https://developers.vtex.com/docs/guides/vtex-io-documentation-promoting-a-workspace-to-master): Optional step to promote the production workspace to master.

>ℹ Check [Deploying a new app version](https://developers.vtex.com/docs/guides/vtex-io-documentation-making-your-new-app-version-publicly-available) for more details on this flow.
```

>ℹ️ To learn about app versioning in VTEX IO, refer to [SemVer standard](https://semver.org/) and our [Understanding app versioning](https://developers.vtex.com/docs/guides/vtex-io-documentation-releasing-a-new-app-version) article.

> Our deployment process integrates an automated Continuous Integration and Continuous Delivery (CI/CD) framework, facilitating seamless building and testing whenever changes occur in the source code.
> 
> Through this CI/CD pipeline, we ensure swift and dependable delivery of new features and updates, while maintaining high standards of code quality.

Reviewer: ```suggestion
Our deployment process integrates an automated Continuous Integration and Continuous Delivery (CI/CD) framework, facilitating seamless building and testing whenever changes occur in the source code. Through this CI/CD pipeline, we ensure swift and dependable delivery of new features and updates, while maintaining high standards of code quality.
```

## Storefront development

VTEX provides different options for storefront development, each with its own characteristics, according to your business needs:

> - [**FastStore**](https://developers.vtex.com/docs/guides/faststore/docs-what-is-faststore): innovative storefront technology based on the [Jamstack](https://jamstack.org/) architecture. This technology prioritizes performance and stability, offering development teams an easily maintainable solution for editing store pages.

Reviewer: ```suggestion
- [**FastStore**](https://developers.vtex.com/docs/guides/faststore/docs-what-is-faststore): Innovative storefront technology based on the [Jamstack](https://jamstack.org/) architecture. FastStore prioritizes performance and stability, offering development teams an easily maintainable solution for editing store pages.
```

> - [**Store Framework**](https://developers.vtex.com/docs/guides/storefront-implementation): emphasizes a low-code, composable commerce model, allowing integration of [VTEX IO apps](https://developers.vtex.com/docs/vtex-io-apps) for dynamic storefronts.

Reviewer: ```suggestion
- [**Store Framework**](https://developers.vtex.com/docs/guides/store-framework): Composable and flexible commerce model, allowing integration of [VTEX IO apps](https://developers.vtex.com/docs/vtex-io-apps) for dynamic storefronts.
```

> - [**Legacy CMS Portal**](https://developers.vtex.com/docs/guides/vtex-io-documentation-migrating-storefront-from-legacy-to-io): VTEX’s inaugural storefront technology, that relies on HTML, CSS, and JavaScript, with code management exclusively through the VTEX Admin. This technology is no longer available to new stores starting with VTEX.

Reviewer: ```suggestion
- [**Legacy CMS Portal**](https://developers.vtex.com/docs/guides/vtex-io-documentation-migrating-storefront-from-legacy-to-io): Legacy storefront technology, that relies on HTML, CSS, and JavaScript, with code management exclusively through the VTEX Admin. This technology is no longer available to new stores starting with VTEX.
```

### Learn more

- [Getting started with storefront solutions](https://developers.vtex.com/docs/guides/getting-started-with-storefront-solutions)
- [FastStore onboarding](https://developers.vtex.com/docs/guides/faststore-onboarding)

> - [Store Framework onboarding](https://developers.vtex.com/docs/guides/store-framework-onboarding)

Reviewer: Since all these guides are from Developer Portal, they should be added as seeAlso in the frontmatter

## Contributions and collaboration

The [VTEX Community](https://community.vtex.com/) is an ecosystem where our clients and partners can interact, ask questions, and exchange information. It is the go-to place to find answers and discuss VTEX products.

The VTEX Community provides forums and discussion groups where members can ask questions, share insights, and seek assistance from peers and experts. These forums cover various topics related to development, implementation, optimization, and best practices for using the VTEX IO platform.

> Developers can participate in the VTEX Community, sharing knowledge, contributing code, and collaborating with other developers to improve the platform and address common challenges and issues.

Reviewer: ```suggestion
Developers can participate in the VTEX Community, sharing knowledge, contributing code, and collaborating with the ecosystem to improve the platform and address common challenges.
```

- [**Publishing**](https://developers.vtex.com/docs/guides/vtex-io-documentation-publishing-an-app): The app version becomes a candidate version, ready for testing and validation in a production workspace.
- [**Testing**](https://developers.vtex.com/docs/guides/ab-tests): Run [A/B tests](#ab-tests).
- [**Deploying**](https://developers.vtex.com/docs/guides/vtex-io-documentation-deploying-the-app-stable-version): Update the app version across all accounts where the app is installed.
- [**Promoting**](https://developers.vtex.com/docs/guides/vtex-io-documentation-promoting-a-workspace-to-master): Optional step to promote the production workspace to master.

>ℹ Check [Deploying a new app version](https://developers.vtex.com/docs/guides/vtex-io-documentation-making-your-new-app-version-publicly-available) for more details on this flow.

>ℹ️ To learn about app versioning in VTEX IO, refer to [SemVer standard](https://semver.org/) and our [Understanding app versioning](https://developers.vtex.com/docs/guides/vtex-io-documentation-releasing-a-new-app-version) article.

> Our deployment process integrates an automated Continuous Integration and Continuous Delivery (CI/CD) framework, facilitating seamless building and testing whenever changes occur in the source code. Through this CI/CD pipeline, we ensure swift and dependable delivery of new features and updates, while maintaining high standards of code quality.

Reviewer: I'm not sure this applies to app development
